compiler/rustc_hir_analysis/src/check/check.rs

```diff
@@ -148,6 +149,47 @@ fn check_union_fields(tcx: TyCtxt<'_>, span: Span, item_def_id: LocalDefId) -> b
     true
 }
 
+/// Check that the unsafe fields do not need dropping.
+fn check_unsafe_fields(tcx: TyCtxt<'_>, span: Span, item_def_id: LocalDefId) {
+    let item_type = tcx.type_of(item_def_id).instantiate_identity();
+    let ty::Adt(def, args) = item_type.kind() else {
+        span_bug!(span, "structs/enums must be ty::Adt, but got {:?}", item_type.kind());
+    };
+    let typing_env = ty::TypingEnv::non_body_analysis(tcx, item_def_id);
+    for variant in def.variants() {
+        for field in &variant.fields {
+            if field.safety != Safety::Unsafe {
+                continue;
+            }
+            let Ok(field_ty) = tcx.try_normalize_erasing_regions(typing_env, field.ty(tcx, args))
+            else {
+                tcx.dcx().span_delayed_bug(span, "could not normalize field type");
+                continue;
+            };
+
+            if !allowed_union_or_unsafe_field(field_ty, tcx, typing_env) {
+                let (field_span, ty_span) = match tcx.hir().get_if_local(field.did) {
+                    // We are currently checking the type this field came from, so it must be local.
+                    Some(Node::Field(field)) => (field.span, field.ty.span),
+                    _ => unreachable!("mir field has to correspond to hir field"),
+                };
+                tcx.dcx().emit_err(errors::InvalidUnsafeField {
+                    field_span,
+                    sugg: errors::InvalidUnsafeFieldSuggestion {
+                        lo: ty_span.shrink_to_lo(),
+                        hi: ty_span.shrink_to_hi(),
+                    },
+                    note: (),
+                });
+            } else if field_ty.needs_drop(tcx, typing_env) {
+                // This should never happen. But we can get here e.g. in case of name resolution errors.
+                tcx.dcx()
+                    .span_delayed_bug(span, "we should never accept maybe-dropping union fields");
```

Comment:
Fixed the first, unsure if fixing the second would actually result in simpler code given the different messages...

Reply:
Well, these are not messages that we expect users to ever see, right? If you deduplicate them, you can just call it a "field" rather than giving it a name. It just needs to be an arbitrary string after all.